python: add functions to facilitate watching job eventlog events

grondo opened the pull request:
This is a WIP stab at adding both synchronous and asynchronous Python functions that use the job eventlogs. This PR adds the following new functions to the `job` module:

 * `job.event_wait(h, jobid, name, eventlog="eventlog")`: Wait for an event named `name` in `eventlog` synchronously. Blocks and returns an `EventLogEvent` for the matched event, or `None` if the eventlog ended with no match.

e.g.
```python
import sys
import flux
from flux import job

jobid = int(sys.argv[1])
name = sys.argv[2]
print(f"waiting for {name}")
print (job.event_wait (flux.Flux(), jobid, name))
```
```console
$ flux python we.py $(flux mini submit sleep 3) 'finish'
1591649139.01139: finish {'status': 0}
```

 * `job.event_watch_async(h, jobid, name="*", eventlog="eventlog")`: returns a JobEventWatchFuture which can be used to asynchronously watch job events via a `.then()` callback. Optionally filter events on the `name` glob (default all events). A JobEventWatchFuture internally queues events for consumption in the continuation with `future.get_event()`. `future.reset()` must be called to "consume" the event.

Example:
```python
import sys
import flux
from flux import job

def cb(future, handle):
    event = future.get_event()
    if event is None:
        # End of stream
        handle.reactor_stop(handle.get_reactor())
        return
    print (event)
    future.reset()

jobid = int(sys.argv[1])
name = sys.argv[2]
h = flux.Flux()
job.event_watch_async(h, jobid, name=name).then(cb, h)

h.reactor_run(h.get_reactor(), 0)
```
```console
$ flux python we.py $(flux mini submit sleep 3) '*'
1591649414.75405: submit {'userid': 1000, 'priority': 16, 'flags': 0}
1591649414.76828: depend {}
1591649414.77174: alloc {'note': 'rank0/core0'}
1591649414.78085: start {}
1591649417.83567: finish {'status': 0}
1591649417.84228: release {'ranks': 'all', 'final': True}
1591649417.84311: free {}
1591649417.84313: clean {}

$ flux python we.py $(flux mini submit sleep 3) 'finish'
1591649450.16344: finish {'status': 0}
```

I initially wanted to build the synchronous function here through the use of the async version, e.g.

```python
future = job.event_watch_async(handle, jobid, name="start")
# block until future fulfilled
future.get_event()
```

However, this doesn't work:
```
Traceback (most recent call last):
  File "we.py", line 9, in <module>
    print(job.event_watch_async(h, jobid, name=name).get_event())
  File "/home/grondo/git/flux-core.git/src/bindings/python/flux/job.py", line 330, in get_event
    raise OSError(ffi.errno, self.error_string())
OSError: [Errno 0] None
```

I *think* this is because the empty `JobWaitEventFuture` returned by `event_watch_async()` returns -1 on the first `.get()`. This might be becuase it doesn't have a way to initialize with the `now` reactor created by the synchronous `.get()`. We may need  a way to register future `init` callbacks from Python. (still looking into it)

Until then, I'm not sure this should be merged. (But maybe the synchronous version is ok)

grondo commented:
This might be nice to get into the next release, as  non-racy alternative to subscribing and parsing the `job-event` events for jobs. Not sure if the current interfaces are acceptable as a start.

SteVwonder commented on src/bindings/python/flux/job.py:
```diff
@@ -267,6 +267,155 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+            self.results.popleft()
+
+
+def event_watch_async(flux_handle, jobid, name="*", eventlog="eventlog"):
```
This is really nice!  For anyone that is familiar with event-driven development and Flux's futures, this is perfect!

For people more familiar with Python than Flux, it might also be nice to have an `event_watch` function that is a generator so that someone can easily write a for loop over the events:

```python
for event in job.event_watch(fh, jobid):
   print("Event: {}".format(event))
```

That loop has the downside of only progressing as fast as the job progresses through its states (that may surprise some users), but I could see how it would be useful (e.g., in `event_wait`).

grondo replied:
The problem is that `event_watch_async` cannot be used synchronously at the moment. See the bottom of the PR description.

This is the right feedback though. I was looking for comments on what interfaces would be most useful.

SteVwonder commented on src/bindings/python/flux/job.py:
```diff
@@ -267,6 +267,155 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+
+    """
+
+    flags = 0
+    result = ffi.new("char *[1]")
+
+    future = Future(RAW.event_watch(flux_handle, int(jobid), eventlog, flags))
+    while True:
+        RAW.event_watch_get(future, result)
+        if result == ffi.NULL:
+            return None
+        event = EventLogEvent(json.loads(ffi.string(result[0]).decode("utf-8")))
+        if event.name == name:
+            return event
+        future.reset()
```
Could this take advantage of `event_watch_async` to reduce code duplication?

```python
return job.event_watch_async(flux_handle, jobid, name=name, eventlog=eventlog).get()
```

EDIT: I initially proposed the following loop until I realized that `event_watch_async` supports the name argument.

```python
for event in job.event_watch(flux_handle, jobid):
    if event.name == name:
        return event
return None
```

grondo replied:
Unfortunately, no. See the above comment. If `event_watch_async().get_event()` is called synchronously it returns `-1`.

grondo commented on src/bindings/python/flux/job.py:
```diff
@@ -267,6 +267,155 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+        """
+        "Initialize from a string or dict eventlog event
+        """
+        self.event = event
```
Oh, yeah, this is probably not necessary, sorry. 

The original event dict was stored for the `__str__` method at first, but then I changed my mind on how that should look.

grondo commented on src/bindings/python/flux/job.py:
```diff
@@ -267,6 +267,155 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+        if self.results:
+            return self.results[0]
+        result = self.pimpl.get(ffi.NULL)
+        if result == -1:
+            raise OSError(ffi.errno, self.error_string())
```
Yeah, I still don't understand it. This happens I think when you call `.get_event()` outside of the reactor. I think the underlying "empty" `flux_future_t` is not initialized. Still trying to figure this out.

grondo commented on src/bindings/python/flux/job.py:
```diff
@@ -804,3 +953,6 @@ def from_command(
         return cls(resources, tasks, attributes=attributes)
+
+
+# vi: ts=4 sw=4 expandtab
```
Oh, sorry about that. I didn't mean to check that line in.

grondo commented:
Thanks for the thorough review @SteVwonder!

I had an idea while going through your comments. If the `name` argument was dropped from `event_watch_async()`, then perhaps the "empty" `flux_future_t` (that is used under `JobEventWatchFuture`) can be dropped. My assumption is that this is the reason `event_watch_async().get_event()` is failing.

The only thing we'd lose is the ability to filter on certain events. However, thinking about it, I'm not sure that is useful enough to warrant the extra complexity.

SteVwonder commented:
> If the name argument was dropped from event_watch_async(), then perhaps the "empty" flux_future_t (that is used under JobEventWatchFuture) can be dropped. My assumption is that this is the reason event_watch_async().get_event() is failing.
> The only thing we'd lose is the ability to filter on certain events. However, thinking about it, I'm not sure that is useful enough to warrant the extra complexity.

:+1: I haven't spent enough time to try and grok the failure here, but I agree that dropping the `name` argument in exchange for a working synchronous version is the right trade-off to make. Presumably we can still provide the `event_wait` function with the `name` argument via filtering implemented within that function?

SteVwonder commented:
Also, sorry for missing your note at the bottom of the original post and the associated noise in the PR comments. :facepalm:

SteVwonder commented:
Just a note that we should remember to ping Joe Koning and the Merlin team once this PR is merged.  They have an interest in using an interface like this.  (If only my TODO manager had a GitHub integration :laughing:)

grondo commented:
Ok, based on excellent feedback from @SteVwonder above, I've just pushed an update (as a new commit) to greatly simplify `job.event_watch_async`, add a `job.event_watch` generator, and then switch `job.event_wait()` to use the generator. I agree this turned out much better.

Examples:

* watch both the job eventlog and exec.eventlog asynchronously
```python
import sys
import flux
from flux import job

jobid = int(sys.argv[1])

def cb(future, h):
    event = future.get_event()
    if event is None:
        return
    if event.name == "start":
        job.event_watch_async(h, jobid, eventlog="guest.exec.eventlog").then(cb, h)
    print(event)
    future.reset()

h = flux.Flux()
job.event_watch_async(h, jobid).then(cb, h)
h.reactor_run()
```
```
$ flux python event-watch.py $(flux mini submit -n4 sleep 1)
1592431436.07016: submit {'userid': 1000, 'priority': 16, 'flags': 0}
1592431436.08510: depend {}
1592431436.08869: alloc {'note': 'rank0/core[0-3]'}
1592431436.09575: start {}
1592431436.08975: init {}
1592431436.09228: starting {}
1592431436.13158: shell.init {'leader-rank': 0, 'size': 1, 'service': '1000-shell-66199673110528'}
1592431436.13997: shell.start {'task-count': 4}
1592431437.14495: complete {'status': 0}
1592431437.14501: cleanup.start {'ranks': 'all'}
1592431437.14687: cleanup.finish {'ranks': 'all', 'rc': 0}
1592431437.14796: done {}
1592431437.14525: finish {'status': 0}
1592431437.15217: release {'ranks': 'all', 'final': True}
1592431437.15313: free {}
1592431437.15315: clean {}
```

 * Use of generator to iterate all events
```python
jobid = int(sys.argv[1])
for event in job.event_watch(flux.Flux(), jobid):
    print(event)
```

 * Use of `job.event_wait()`:
```python
import base64
import sys
import flux
from flux import job

jobid = int(sys.argv[1])

h = flux.Flux()

# block until "start" event, so we know exec.eventlog is available
job.event_wait(h, jobid, "start")

# block until "shell.init" event, so we know output eventlog is available
job.event_wait(h, jobid, "shell.init", eventlog="guest.exec.eventlog")

# process output eventlog (doesn't differentiate between stdout/err)
for event in job.event_watch(h, jobid, eventlog="guest.output"):
    if event.name == "data" and "data" in event.context:
        print(base64.b64decode(event.context["data"]).decode("utf-8").strip())
```

```
$ flux python event-wait.py $(flux mini submit -n4 hostname)
asp
asp
asp
asp
```

grondo commented:
If the interface looks ok, I can squash away my previous mistakes and write some tests.

grondo commented:
One thing: it is very easy to forget the `future.reset()` call in the asynchronous version.

I wonder if, for python, it would be better if the future was auto-reset after a call to `event = future.get_event()`. The downside would be that a second call to `get_event()` would block (or return the next result instead of the current one).

grondo commented:
Oh, the other thought was how to return "no matching event" from the `job.event_wait()` call. For now, it throws `OSError(errno.ENODATA, "eventlog ended before event={name}")`. I'm open to other ideas.

grondo commented:
Also, there is the matter of canceling the streaming `flux_job_watch_event(3)` RPC.
Hopefully this will automatically do that when the `JobEventWatchFuture` is garbage collected?
```diff
diff --git a/src/bindings/python/flux/job.py b/src/bindings/python/flux/job.py
index 44c5a340d..835434262 100644
--- a/src/bindings/python/flux/job.py
+++ b/src/bindings/python/flux/job.py
@@ -360,18 +360,34 @@ class JobEventWatchFuture(Future):
     Adds get_event() method to return an EventLogEntry event
     """
 
+    def __del__(self):
+        if self.needs_cancel is not False:
+            self.cancel()
+
+    def __init__(self, future_handle):
+        super().__init__(future_handle)
+        self.needs_cancel = True
+
     def get_event(self):
         result = ffi.new("char *[1]")
         try:
             RAW.event_watch_get(self.pimpl, result)
         except OSError as exc:
             if exc.errno == errno.ENODATA:
+                self.needs_cancel = False
                 return None
             else:
                 # re-raise all other exceptions
                 raise
         return EventLogEvent(ffi.string(result[0]).decode("utf-8"))
 
+    def cancel(self):
+        """Cancel a streaming job.event_watch_async() future
+        """
+        RAW.event_watch_cancel(self.pimpl)
+        self.needs_cancel = False
+
+

```

Things seem to work without this, but I worry because `job.event_wait()` and `job.event_watch()` will definitely leave the RPC un-canceled. (But I can't remember what, if anything, we have that cleans up after this case in the background)

SteVwonder commented on src/bindings/python/flux/job.py:
```diff
@@ -411,63 +394,53 @@ def event_watch_async(flux_handle, jobid, name="*", eventlog="eventlog"):
-    Future(RAW.event_watch(flux_handle, int(jobid), eventlog, flags)).then(
-        event_watch_cb
-    )
+def event_watch(flux_handle, jobid, eventlog="eventlog", flags=0):
```
Whatever we decided w.r.t. `flags` for in `event_watch_async` should apply here too.

SteVwonder commented:
> Also, there is the matter of canceling the streaming flux_job_watch_event(3) RPC. Hopefully this will automatically do that when the JobEventWatchFuture is garbage collected?

The `Future` class (and all of it's descendants) have `flux_future_destroy` called on them in the `Future.__del__` by default.  Does `flux_future_destroy` take care of canceling futures?  If not, then what you have looks fine (just be sure to call `super().__del__` at the end of the `__del__` method that you'll be adding).

Once we have another streaming service to work with (maybe we already have one), maybe can abstract that functionality up into a `StreamingRPC` or `StreamingFuture` class.  Just a thought for later down the road.

SteVwonder commented:
> Does flux_future_destroy take care of canceling futures?

https://github.com/flux-framework/flux-core/issues/688 makes it sound like a matchtag is leaked when an uncanceled rpc is destroyed.  It's a 4 year old issue, so I wonder if it is still the case.  @garlick?

grondo commented:
> (just be sure to call super().__del__ at the end of the __del__ method that you'll be adding).

Thanks! I'll add that to the commits I just pushed.

BTW, just pushed a set of tests as well.

grondo commented on src/bindings/python/flux/job.py:
```diff
@@ -357,40 +356,24 @@ def context(self):
 
 
-def event_watch_async(flux_handle, jobid, name="*", eventlog="eventlog"):
+def event_watch_async(flux_handle, jobid, eventlog="eventlog", flags=0):
```
Ok, fine with me!

grondo commented:
> > For now, it throws OSError(errno.ENODATA, "eventlog ended before event={name}"). I'm open to other ideas.

> One alternative would be to just return None.

I thought about this, but this means that every call to wait_event has to check its return. I think the interface is much nicer if an exception is thrown in the case where the eventlog ended without a matching event, since most of the time this is an error condition.

And in fact, it might lead to buggy code if the event appears 99 times out of 100, a user might discard the return value like:

```python
  job.wait_event(h, jobid, "start")
  # Do something that depends on start event having been reached
```

In the strange case where a start event is not reached (e.g. cancellation), the code will proceed.

grondo commented:
Ok, @SteVwonder, I pushed some fixup commits addressing your comments. Thanks for taking a look!

I like your idea of a "autoreset=False" flag for the `.get_event()` method. I'll add something like that and corresponding tests later.

Hoping to get this feature into our next tag.

garlick commented:
>> Does flux_future_destroy take care of canceling futures?
>>
>#688 makes it sound like a matchtag is leaked when an uncanceled rpc is destroyed. It's a 4 year old issue, so I wonder if it is still the case. @garlick?

The matchtag is leaked when the future is destroyed - however, it is eventually retired if a terminating response is received for it after the fact.  There is no automatic sending of any kind of cancellation message on destroy because we don't have a standard mechanism for that.

grondo commented:
Here's another idea: Should `job.event_wait()` (optionally) raise a Python exception if there is a sev 0 job exception while waiting for a different event?

We could add a `JobException` class that acts as a container for the exception event, and throw this if `event.name == "exception"`. That would allow this to work:

```python
try:
    job.event_wait(h, jobid, "start")
except job.JobException as e:
    print (f"{e.timestamp:.2f}: job.exception type={e.type} note={e.note}")
    sys.exit(1)
```

```console
$ flux python event-wait.py $(flux mini submit -n 8 hostname)
1592449408.19: job.exception type=alloc note=unsatisfiable request
```

I'll push an example here, which I'm happy to remove if it is a step too far for this PR.

grondo commented:
Ok, removed WIP as I think this is ready.

I did add the `autoreset` flag to `JobEventWatchFuture`. I think that turned out nicely -- good suggestion @SteVwonder.

I also pushed a commit that raises a `JobException` when an `exception` event is posted in the eventlog before the matching event in `job.event_wait()` (unless the matching event is literally `exception`). I'd be fine dropping the commit for now though if we aren't quite sure this is right direction.

I've been having trouble getting this PR to get through address-sanitizer GH check without hanging or timing out, though. However, the only affected test - `t0010-job.py` is completing, so not sure what is going on there.

grondo commented:
Thanks! We can fiddle with the interfaces later if we need to do so.

SteVwonder commented:
Thanks @grondo!  This turned out really nice!  Thanks for getting this in before the tag.
